odl/solvers/functional/functional.py

```diff
@@ -0,0 +1,887 @@
+
+            If ``other`` is a `Functional`, ``sum`` is a `FunctionalSum`.
+
+        See Also
+        --------
+        Operator.__add__ : implementation of __add__ for Operators.
+        """
+        if other in self.domain.field:
+            return FunctionalScalarSum(self, other)
+        elif isinstance(other, Functional):
+            return FunctionalSum(self, other)
+        else:
+            return super().__add__(other)
+
+    # Since addition is commutative, right and left addition is the same
+    __radd__ = __add__
+
+    def __sub__(self, other):
+        """Return ``self - other``."""
+        return self.__add__(-1 * other)
+
+
+class FunctionalLeftScalarMult(Functional, OperatorLeftScalarMult):
+
+    """Scalar multiplication of functional from the left.
+
+    Given a functional ``f`` and a scalar ``scalar``, this represents the
+    functional
+
+        ``(scalar * f)(x) == scalar * f(x)``.
+    """
+
+    def __init__(self, func, scalar):
+        """Initialize a new instance.
+
+        Parameters
+        ----------
+        func : `Functional`
+            Functional to be scaled.
+        scalar : `float`
+            Number with which to scale the functional.
+        """
+        if not isinstance(func, Functional):
+            raise TypeError('`func` {!r} is not a `Functional` instance.'
+                            ''.format(func))
+
+        self.__scalar = func.range.element(scalar)
+
+        grad_lipschitz = (0 if scalar == 0
+                          else np.abs(scalar) * func.grad_lipschitz)
+        Functional.__init__(self, space=func.domain,
+                            linear=func.is_linear or scalar == 0,
+                            grad_lipschitz=grad_lipschitz)
+
+        OperatorLeftScalarMult.__init__(self, operator=func, scalar=scalar)
+
+    @property
+    def scalar(self):
+        return self.__scalar
+
+    @property
+    def functional(self):
+        """The original functional."""
+        return self.operator
+
+    @property
+    def gradient(self):
+        """Gradient operator of the functional."""
+        return self.scalar * self.functional.gradient
+
+    @property
+    def convex_conj(self):
+        """Convex conjugate functional of the scaled functional."""
+        # The helper function only allows positive scaling parameters.
+        # Otherwise it gives an error.
+
+        if self.scalar <= 0:
+            raise ValueError('Scaling with nonpositive values have no convex '
+                             'conjugate. Current value: {}.'
+                             ''.format(self.scalar))
+
+        return self.scalar * self.functional.convex_conj * (1.0 / self.scalar)
+
+    @property
+    def proximal(self):
+        """Proximal factory of the scaled functional.
+
+        See Also
+        --------
+        proximal_zero
+        """
+
+        if self.scalar < 0:
+            raise ValueError('Proximal operator of functional scaled with a '
```

Comment:
Our exception message style is to not capitalize the first letter or use punctuation unless the message is more than one sentence.